### PR TITLE
Fix history and logbook for input_datetime

### LINKED ISSUE #7309: input_datetime objects' new History tab shows incorrect entries
## Checklist

- [x] I have updated to the latest available Home Assistant version.
- [x] I have cleared the cache of my browser.
- [x] I have tried a different browser to see if it is related to my browser.

## The problem

Opening a 'helper' (e.g. an input_datetime)'s more info dialog ([this new feature introduced in 0.115.x](https://www.home-assistant.io/blog/2020/09/17/release-115/#updated-more-info-dialog)) and clicking the History tab shows the full history of when each change was made, as expected. However, each value is the current value instead of the historic one, e.g. if I just changed an input_datetime from 06:00 to 07:00, the history will show _all_ previous times as 07:00. 

## Expected behavior

History tab should show the correct historical values.

## Steps to reproduce

1. Create a helper (mine were created in yml rather than via the UI, not sure if this makes a difference.)
2. Navigate to Lovelace
3. Add the helper to Lovelace.
4. Edit the helper's values a few times
5. Click on more info for the helper.
6. Click on the history tab
7. See that all historic values are the same.

## Environment

- Home Assistant release with the issue: 0.116.x
- Last working Home Assistant release (if known): N/A - probably didn't work in 0.115.x either, feature introduced then.
- Browser and browser version: Firefox latest
- Operating system: MacOS Catalina latest

## State of relevant entities

```yaml

```

## Problem-relevant configuration

```yaml

```

## Javascript errors shown in your browser console/inspector

```txt

```

### PR DESCRIPTION
**Pull request by spacegaier**
Probably obsolete once core PR https://github.com/home-assistant/core/pull/45208 has been merged.

## Proposed change

The underlying reason for the problem that both history and logbook for `input_datetime` were incorrect, stems from the following:

Function `computeStateDisplay()` assumed that for `input_datetime` the `attributes` of the state would always be available, both to determine the state type ("date", "time" or "datetime" via the `hass_time` and `hass_date` flags) as well as for the actual values. For separate reasons that however was never true:

History: The backend data retrieval always used "minimal_response" which meant only the first and last returned state contained the attributes.
Logbook: The backend webservice never provided the state attributes at all.

The solution I came up with consists of the following parts:
1. For the history webservice call check if we are requesting data for an `input_datetime`. If yes, never set the `minimal_response` parameter.
2. Adjust `computeStateDisplay()` so that for the actual state display, we no longer use the attributes, but the state itself. We still use the attributes to determine the type ("date", "time" or "datetime"). Two scenarios come into play here:
2.1 Execution for history: Thanks to point 1, we will always have the attributes of the history state entries now => so even if in the middle of the historic data the type changed, the display will be correct since each state carries the type info ("date", "time" or "datetime").
2.2. Execution for logbook: Since we will never have the historic state attributes here, we have to rely on the type provided by the current state attributes. That means that if in the past the entity was of type "date" and now has been switched to "time", some logbook entries will show incorrect data, since we always have to assume "time" now.

If we are fine with the restriction from the logbook logic (and I cannot imagine a scenario where an `input_datetime` is changing its type), then we could even revert the changes in regards to "minimal_response" for the history to reduce the data load.

Note: The initial history panel display will request historic data for all entities => no entity filter => cannot check for `input_datetime` => behavior the same as int he logbook (we have to assume the type based on the current attributes).

## Type of change

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes https://github.com/home-assistant/frontend/issues/7309
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/common/entity/compute_state_display.ts [hunk 1]
```diff
@@ -30,34 +30,19 @@ export const computeStateDisplay = (
   if (domain === "input_datetime") {
     let date: Date;
     if (!stateObj.attributes.has_time) {
-      date = new Date(
-        stateObj.attributes.year,
-        stateObj.attributes.month - 1,
-        stateObj.attributes.day
-      );
+      date = new Date(compareState);
```
Comment (bramkragten) [inline src/common/entity/compute_state_display.ts:33]: What is the format of the state?

Comment (bramkragten) [reply src/common/entity/compute_state_display.ts:33]: If it is not an ISO datetime string, this will not work in some browsers, like Safari.

Comment (spacegaier) [reply src/common/entity/compute_state_display.ts]: Formats of `input_datetime`:
Date: YYYY-MM-DD
Time: HH:MM:SS
Datetime: YYYY-MM-DD HH:MM:SS

Reg. Safari: I had taken that into account for datetime, but not for date. I now replaced the Date() constructor / date parsing, so it should be safe now.

File: src/data/cached-history.ts [hunk 2]
```diff
@@ -121,6 +121,10 @@ export const getRecentWithCache = (
 
   const curCacheProm = cache.prom;
 
+  // For accurate history display of an input_datetime we require the attributes
+  // => we need a complete response => minimalResponse = false
+  const minimalResponse = entityId.split(".")[0] !== "input_datetime";
```
Comment (bramkragten) [inline src/data/cached-history.ts:126]: How would we do this if we are getting the history of multiple entities?

It is actually handled in the backend:
https://github.com/home-assistant/core/blob/dev/homeassistant/components/history/__init__.py#L71-L76

**Comment by bramkragten** [comment]
> 2.2. Execution for logbook: Since we will never have the historic state attributes here, we have to rely on the type provided by the current state attributes. That means that if in the past the entity was of type "date" and now has been switched to "time", some logbook entries will show incorrect data, since we always have to assume "time" now.

It will actually error now, as it expects a certain format for the state with those attributes.

File: src/data/cached-history.ts (hunk 2 above)
Comment (spacegaier) [reply src/data/cached-history.ts:126]: Good catch. Core PR is on its way. Then from the looks of it, this frontend PR can be completely rejected (and I could have saved myself a lot of time, had I seen the backend exclusion list).

**Comment by spacegaier** [comment]
Should be fixed by the alternative core PR: https://github.com/home-assistant/core/pull/45208

**Pull request closed without merging**